Meetup 25 Resources (#190)
w3c/wot-cg: Meetups/25/2025-03-14-minutes.md

- Tatiana
- Daniel Peintner
- Alexandru Sorici
- Terry Payne
- Julian Padget
- Stefan Bußweiler
- Raul
- Patrick
- Lorenzo Gigli
- Gianluca Aguzzi
- Alessandro Giugno

:writing_hand: **Scribe:** Cris

----

### Introduction

- Ege: welcome everyone! This is a joint call between WoT, WebAgents and Web Thing Protocol CG
- Ege: Working at siemens and co-lead of CG group
- Cris: Work in standards and in a company called Archeion.
- Andrei: Web agents CG
- Rem: Professor working with multi agents system, involved in hypermedia multi agents systems
- Ben: Founder of Krellian, Web Thing Protocol CG would ask anyone who is interested in the development of the protocol to join the group.

### News

Ege:two weeks from now we will host a new meetup about HiveOT

Ege: More are planned

Ege: if you anything to present let us know!

Andrei: Quick news from us, we are organizing a workshop. If you have research or innovation projects please reach out.

Ben: We have started writing the specification of Web Thing Protocol and there is a strawman proposal (watch out is getting out of date).

Ege: Joining the groups is free, you can join using the button.

### Meetup Presentation

> Ege: Today we have Robert Winkler one of the Driver of LMOS project. He is focused on internet of Things and conversational AI. Solution architect at day and a gamer at night.
> 
> Robert: We will talk about LMOS at Eclipse but also how we used WoT in that system

Reviewer: ```suggestion
Ege: Today we have Robert Winkler, one of the drivers of the LMOS project. He is focused on Internet of Things and conversational AI. Solution architect by day and a gamer by night.

Robert: We will talk about LMOS at Eclipse as well as how we used WoT in that system.
```

> Robert: At DT we are bring GEnAi to solve sales and services. We are going soon to integrate voice in the system. That solution is based on a LLM. You can augment LLMs with tools, usually functions to performing actions in the digital/physical world. The agent can also access short and long term memory.
> 
> Robert: the focus of today is more about the process we went trough when building such system and where WoT played a crucial role.

Reviewer: ```suggestion
Robert: At DT, we are using GEnAi to solve sales and services. We will soon be integrating voice into the system. That solution is based on a LLM. You can augment LLMs with tools, usually functions to performing actions in the digital/physical world. The agent can also access short and long term memory.

Robert: Today's focus is more about the process we went trough when building such a system and where WoT played a crucial role.
```

> Robert: We split the problem in different agents to improve performance (one agent was performing bad).
> 
> Robert: Now we don't allow communication between agents, main challenge here is to avoid loops. At DT we are switching from API to agents. Soon we'll have an internet of agents not API calls.

Reviewer: ```suggestion
Robert: We split the problem over different agents to improve performance (one agent was performing poorly).

Robert: Now we don't allow communication between agents; main challenge is to avoid loops. At DT, we are switching from API to agents. Soon we'll have an internet of agents, not API calls.
```

> Robert: But then new challenges unlocks: discovery, protocol, security, interoperability, group, metadata, versioning. A lot of problems that need to be solved.
> 
> Robert: At Eclipse LMOS we are building on top of  existing solutions to solve the challenges above. We use: Discovery Specification, Web Thing protocol, and Security.

Reviewer: ```suggestion
Robert: But then new challenges are unlocked: discovery, protocol, security, interoperability, group, metadata, versioning. A lot of problems that need to be solved.

Robert: At Eclipse LMOS, we are building on top of existing solutions to solve the challenges above. We use Discovery Specification, Web Thing protocol, and Security.
```

> Robert: When we implement an agent we describe it using a Thing Description. From the consumer prospective can talk with the agent with different protocol. Agents can perform long running tasks. We need messaging protocols. We can use Websockets from some agents but also HTTP. Agents can register themselves to a directory (but also tools which are described by TDs too).
> 
> Robert: LMOS is built around JVM (kotlin) but soon we will have Python runtime to use the well-know AI tooling ecosystem.

Reviewer: ```suggestion
Robert: When we implement an agent, we describe it using a Thing Description. From the consumer prospective, we can talk with the agent with different protocols. Agents can perform long running tasks. We need messaging protocols. We can use Websockets and/or HTTP from some agents. Agents (as well as tools which are described by TDs) can register themselves to a directory.

Robert: LMOS is built around JVM (`kotlin`) but soon we will have Python runtime to use the well-know AI tooling ecosystem.
```

> Robert: you can plug and play protocols. you can use DID for identifying a tool or agent. With the DID yuo can discovery TDs.
> 
> Robert: we develop a DSL for developing agents which use our runtime for WoT written in Kotlin.

Reviewer: ```suggestion
Robert: You can plug and play protocols. You can use DIDs for identifying tools and/or agents. With the DID, you can discovery TDs.

Robert: We developed a DSL for developing agents which use our runtime for WoT written in Kotlin.
```

> Robert: The td is composed by Agent Metadata, Agent API (affordances: Events, Actions, Properties), and, Agent Capabilities.
> 
> Robert: From security we are using WoT security schemas. At DT we use a combination M-to-M token and User token.

Reviewer: ```suggestion
Robert: The TD is composed of Agent Metadata, Agent API (affordances: Events, Actions, Properties), and Agent Capabilities.

Robert: From security, we are using WoT security schemas. At DT, we use a combination M-to-M token and User token.
```

Robert: In spring boot you can write an agent using annotations. As dev I don't care about the protocol. The TD is generated by the protocol. It's a code first framework.

Robert: we are using Thingweb for runtime and Webthings for gateway control. I converted the things in the testing Webthings gateway to Agent Tools. I can ask a questions and see the result of the actions.

> Shows a demo

> Robert: I just used the description and I was able to create a scenario. We open-sourced the whole LMOS platform.
> 
> Robert: We defined Capabilities to augment Agent description in TD. Operator can fetch Agent Capabilities. We have build a complex CI/CD flow specific for agents, we support canary deployments.
> 
> Robert: We defined the concept of channels to route the right query to the right agent. The entrypoint is Supervisor Agent that talks with LMOS operator which is capable of creating channel information.

Reviewer: ```suggestion
Robert: I just used the description and I was able to create a scenario. We open-sourced the whole LMOS platform.

Robert: We defined Capabilities to augment Agent descriptions in TD. Operators can fetch Agent Capabilities. We have built a complex CI/CD flow specific for agents. We support canary deployments.

Robert: We defined the concept of channels to route the right query to the right agent. The entrypoint is Supervisor Agent that talks with LMOS operator which is capable of creating channel information.
```

> Shows a demo

> Robert: if you don't have the agent capable of handling a request the query the system respond accordingly, but you can later install it.
> 
> Robert: Additionally, the system automatically provide different context information: user and system.
> 
> Robert: the whole project is a WIP let's collaborate to make it real.

Reviewer: ```suggestion
Robert: If you don't have an agent capable of handling a request or query, the system responds accordingly, but you can install it later.

Robert: Additionally, the system automatically provides different context information: user and system.

Robert: The whole project is a WIP. Let's collaborate to make it real!
```

### Discussion

> Ege: thank you for the presentation and thank you for making it open source.
> 
> Samuele: Are the use cases part of the standard or custom for agents.
> 
> Robert: They are not standard. They are made for Arc (Agent DSL).
> 
> Samuele: It reminded my MCP for prompts.

Reviewer: ```suggestion
Ege: Thank you for the presentation and thank you for making it open source.

Samuele: Are the use cases part of the standard or custom for agents?

Robert: They are not standard. They are made for Arc (Agent DSL).

Samuele: It reminded me of MCP for prompts.
```

> Ben: Really cool project, great to see to use  Web Thing protocol. I'm curious about agents interacting with devices using TDs. The interesting part is Agent talk with each other, in particular using natural language. I show a demo using a particular audio protocol (tones instead of natural language).
> 
> Robert: Quite interesting demo, we are looking in M to M too. There might be more performant protocol than audio. The protocol negotiation is very important. We can re-use existing solutions and I like the meta idea of WoT, protocol and media type negotiation is very interesting. P2P for WoT is also interesting. There is a new agency formed to build a new protocol... I fear that we are re-inventing the wheel. Is important we collaborate and build something together.

Reviewer: ```suggestion
Ben: Really cool project, great to see use of Web Thing protocol. I'm curious about agents interacting with devices using TDs. The interesting part is Agents talking with each other, in particular using natural language. I show a demo using a particular audio protocol (tones instead of natural language).

Robert: Quite interesting demo, we are looking into M-to-M too. There might be more performant protocols than audio. The protocol negotiation is very important. We can re-use existing solutions, and I like the meta idea of WoT. Protocol and media type negotiation is very interesting. P2P for WoT is also interesting. There is a new agency formed to build a new protocol... I fear that we are reinventing the wheel. It's important that we collaborate and build something together.
```

Samuele: We worked on context switching <https://arxiv.org/abs/2410.11905>

Sam: Is there some platform to publish agents? Like pip or npm?

> Robert: In MCP there are multiple registries. We will built something similar on top of WoT. We can start working on it in DT, but later make it public. Another interesting point is micro transactions between agents.

Reviewer: ```suggestion
Robert: In MCP, there are multiple registries. We will build something similar on top of WoT. We can start working on it in DT, and later make it public. Another interesting point is micro-transactions between agents.
```

> Jonathan: How to know which is the best agents when getting a query?

Reviewer: ```suggestion
Jonathan: How do you know which is the best agent when getting a query?
```

Robert: It is challenging.

> Jonathan: Could we end up to have something similar to SEO?
> 
> Robert: That's good point, it might be a possible scenario.
> 
> Jonathan: Does it make sense to use metadata to pick the right agent.

Reviewer: ```suggestion
Jonathan: Could we end up having something similar to SEO?

Robert: That's good point. It might be a possible scenario.

Jonathan: Does it make sense to use metadata to pick the right agent?
```

> Rem: Interesting point of inter agent communication.
> 
> Robert: In Matter they publishing metadata in blockchain.

Reviewer: ```suggestion
Rem: Interesting point of inter-agent communication.

Robert: In Matter, they publish metadata to a blockchain.
```

> Cris: how can I choose an agent?
> 
> Robert: Apps will die, and you can install agents in your systems.
> 
> Cris: then you can choose the right one

Reviewer: ```suggestion
Cris: How can I choose an agent?

Robert: Apps will die, and you can install agents in your systems.

Cris: Then you can choose the right one.
```

> Robert: yeah, and then you can use bitcoin to pay for usage.
> 
> Sam: Getting new packages in this platform is not necessary to sign, we can do it later.
> 
> Robert: yes, I totally agree. MCP at the beginning didn't even have security.
> 

Reviewer: ```suggestion
Robert: Yeah, and then you can use Bitcoin to pay for usage.

Sam: Getting new packages in this platform is not necessary to sign; we can do it later.

Robert: Yes, I totally agree. At the beginning, MCP didn't even have security.

```
